File: src/main/java/io/github/erp/service/WeeklyCounterfeitHoldingService.java
package io.github.erp.service;

/*-
 * Erp System - Mark VIII No 2 (Hilkiah Series) Server ver 1.6.1
 * Copyright © 2021 - 2023 Edwin Njeru (mailnjeru@gmail.com)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import io.github.erp.service.dto.WeeklyCounterfeitHoldingDTO;
import java.util.Optional;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * Service Interface for managing {@link io.github.erp.domain.WeeklyCounterfeitHolding}.
 */
public interface WeeklyCounterfeitHoldingService {
    /**
     * Save a weeklyCounterfeitHolding.
     *
     * @param weeklyCounterfeitHoldingDTO the entity to save.
     * @return the persisted entity.
     */
    WeeklyCounterfeitHoldingDTO save(WeeklyCounterfeitHoldingDTO weeklyCounterfeitHoldingDTO);

    /**
     * Partially updates a weeklyCounterfeitHolding.
     *
     * @param weeklyCounterfeitHoldingDTO the entity to update partially.
     * @return the persisted entity.
     */
    Optional<WeeklyCounterfeitHoldingDTO> partialUpdate(WeeklyCounterfeitHoldingDTO weeklyCounterfeitHoldingDTO);

    /**
     * Get all the weeklyCounterfeitHoldings.
     *
     * @param pageable the pagination information.
     * @return the list of entities.
     */
    Page<WeeklyCounterfeitHoldingDTO> findAll(Pageable pageable);

    /**
     * Get the "id" weeklyCounterfeitHolding.
     *
     * @param id the id of the entity.
     * @return the entity.
     */
    Optional<WeeklyCounterfeitHoldingDTO> findOne(Long id);

    /**
     * Delete the "id" weeklyCounterfeitHolding.
     *
     * @param id the id of the entity.
     */
    void delete(Long id);

    /**
     * Search for the weeklyCounterfeitHolding corresponding to the query.
     *
     * @param query the query of the search.
     *
     * @param pageable the pagination information.
     * @return the list of entities.
     */
    Page<WeeklyCounterfeitHoldingDTO> search(String query, Pageable pageable);
}

File: erp-client/src/main/webapp/app/entities/gdi-data/weekly-counterfeit-holding/service/weekly-counterfeit-holding.service.ts
///
/// Erp System - Mark X No 11 (Jehoiada Series) Client 1.7.8
/// Copyright © 2021 - 2024 Edwin Njeru (mailnjeru@gmail.com)
///
/// This program is free software: you can redistribute it and/or modify
/// it under the terms of the GNU General Public License as published by
/// the Free Software Foundation, either version 3 of the License, or
/// (at your option) any later version.
///
/// This program is distributed in the hope that it will be useful,
/// but WITHOUT ANY WARRANTY; without even the implied warranty of
/// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
/// GNU General Public License for more details.
///
/// You should have received a copy of the GNU General Public License
/// along with this program. If not, see <http://www.gnu.org/licenses/>.
///

import { Injectable } from '@angular/core';
import { HttpClient, HttpResponse } from '@angular/common/http';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import * as dayjs from 'dayjs';

import { isPresent } from 'app/core/util/operators';
import { DATE_FORMAT } from 'app/config/input.constants';
import { ApplicationConfigService } from 'app/core/config/application-config.service';
import { createRequestOption } from 'app/core/request/request-util';
import { SearchWithPagination } from 'app/core/request/request.model';
import { IWeeklyCounterfeitHolding, getWeeklyCounterfeitHoldingIdentifier } from '../weekly-counterfeit-holding.model';

export type EntityResponseType = HttpResponse<IWeeklyCounterfeitHolding>;
export type EntityArrayResponseType = HttpResponse<IWeeklyCounterfeitHolding[]>;

@Injectable({ providedIn: 'root' })
export class WeeklyCounterfeitHoldingService {
  protected resourceUrl = this.applicationConfigService.getEndpointFor('api/weekly-counterfeit-holdings');
  protected resourceSearchUrl = this.applicationConfigService.getEndpointFor('api/_search/weekly-counterfeit-holdings');

  constructor(protected http: HttpClient, protected applicationConfigService: ApplicationConfigService) {}

  create(weeklyCounterfeitHolding: IWeeklyCounterfeitHolding): Observable<EntityResponseType> {
    const copy = this.convertDateFromClient(weeklyCounterfeitHolding);
    return this.http
      .post<IWeeklyCounterfeitHolding>(this.resourceUrl, copy, { observe: 'response' })
      .pipe(map((res: EntityResponseType) => this.convertDateFromServer(res)));
  }

  update(weeklyCounterfeitHolding: IWeeklyCounterfeitHolding): Observable<EntityResponseType> {
    const copy = this.convertDateFromClient(weeklyCounterfeitHolding);
    return this.http
      .put<IWeeklyCounterfeitHolding>(
        `${this.resourceUrl}/${getWeeklyCounterfeitHoldingIdentifier(weeklyCounterfeitHolding) as number}`,
        copy,
        { observe: 'response' }
      )
      .pipe(map((res: EntityResponseType) => this.convertDateFromServer(res)));
  }

  partialUpdate(weeklyCounterfeitHolding: IWeeklyCounterfeitHolding): Observable<EntityResponseType> {
    const copy = this.convertDateFromClient(weeklyCounterfeitHolding);
    return this.http
      .patch<IWeeklyCounterfeitHolding>(
        `${this.resourceUrl}/${getWeeklyCounterfeitHoldingIdentifier(weeklyCounterfeitHolding) as number}`,
        copy,
        { observe: 'response' }
      )
      .pipe(map((res: EntityResponseType) => this.convertDateFromServer(res)));
  }

  find(id: number): Observable<EntityResponseType> {
    return this.http
      .get<IWeeklyCounterfeitHolding>(`${this.resourceUrl}/${id}`, { observe: 'response' })
      .pipe(map((res: EntityResponseType) => this.convertDateFromServer(res)));
  }

  query(req?: any): Observable<EntityArrayResponseType> {
    const options = createRequestOption(req);
    return this.http
      .get<IWeeklyCounterfeitHolding[]>(this.resourceUrl, { params: options, observe: 'response' })
      .pipe(map((res: EntityArrayResponseType) => this.convertDateArrayFromServer(res)));
  }

  delete(id: number): Observable<HttpResponse<{}>> {
    return this.http.delete(`${this.resourceUrl}/${id}`, { observe: 'response' });
  }

  search(req: SearchWithPagination): Observable<EntityArrayResponseType> {
    const options = createRequestOption(req);
    return this.http
      .get<IWeeklyCounterfeitHolding[]>(this.resourceSearchUrl, { params: options, observe: 'response' })
      .pipe(map((res: EntityArrayResponseType) => this.convertDateArrayFromServer(res)));
  }

  addWeeklyCounterfeitHoldingToCollectionIfMissing(
    weeklyCounterfeitHoldingCollection: IWeeklyCounterfeitHolding[],
    ...weeklyCounterfeitHoldingsToCheck: (IWeeklyCounterfeitHolding | null | undefined)[]
  ): IWeeklyCounterfeitHolding[] {
    const weeklyCounterfeitHoldings: IWeeklyCounterfeitHolding[] = weeklyCounterfeitHoldingsToCheck.filter(isPresent);
    if (weeklyCounterfeitHoldings.length > 0) {
      const weeklyCounterfeitHoldingCollectionIdentifiers = weeklyCounterfeitHoldingCollection.map(
        weeklyCounterfeitHoldingItem => getWeeklyCounterfeitHoldingIdentifier(weeklyCounterfeitHoldingItem)!
      );
      const weeklyCounterfeitHoldingsToAdd = weeklyCounterfeitHoldings.filter(weeklyCounterfeitHoldingItem => {
        const weeklyCounterfeitHoldingIdentifier = getWeeklyCounterfeitHoldingIdentifier(weeklyCounterfeitHoldingItem);
        if (
          weeklyCounterfeitHoldingIdentifier == null ||
          weeklyCounterfeitHoldingCollectionIdentifiers.includes(weeklyCounterfeitHoldingIdentifier)
        ) {
          return false;
        }
        weeklyCounterfeitHoldingCollectionIdentifiers.push(weeklyCounterfeitHoldingIdentifier);
        return true;
      });
      return [...weeklyCounterfeitHoldingsToAdd, ...weeklyCounterfeitHoldingCollection];
    }
    return weeklyCounterfeitHoldingCollection;
  }

  protected convertDateFromClient(weeklyCounterfeitHolding: IWeeklyCounterfeitHolding): IWeeklyCounterfeitHolding {
    return Object.assign({}, weeklyCounterfeitHolding, {
      reportingDate: weeklyCounterfeitHolding.reportingDate?.isValid()
        ? weeklyCounterfeitHolding.reportingDate.format(DATE_FORMAT)
        : undefined,
      dateConfiscated: weeklyCounterfeitHolding.dateConfiscated?.isValid()
        ? weeklyCounterfeitHolding.dateConfiscated.format(DATE_FORMAT)
        : undefined,
      dateSubmittedToCBK: weeklyCounterfeitHolding.dateSubmittedToCBK?.isValid()
        ? weeklyCounterfeitHolding.dateSubmittedToCBK.format(DATE_FORMAT)
        : undefined,
    });
  }

  protected convertDateFromServer(res: EntityResponseType): EntityResponseType {
    if (res.body) {
      res.body.reportingDate = res.body.reportingDate ? dayjs(res.body.reportingDate) : undefined;
      res.body.dateConfiscated = res.body.dateConfiscated ? dayjs(res.body.dateConfiscated) : undefined;
      res.body.dateSubmittedToCBK = res.body.dateSubmittedToCBK ? dayjs(res.body.dateSubmittedToCBK) : undefined;
    }
    return res;
  }

  protected convertDateArrayFromServer(res: EntityArrayResponseType): EntityArrayResponseType {
    if (res.body) {
      res.body.forEach((weeklyCounterfeitHolding: IWeeklyCounterfeitHolding) => {
        weeklyCounterfeitHolding.reportingDate = weeklyCounterfeitHolding.reportingDate
          ? dayjs(weeklyCounterfeitHolding.reportingDate)
          : undefined;
        weeklyCounterfeitHolding.dateConfiscated = weeklyCounterfeitHolding.dateConfiscated
          ? dayjs(weeklyCounterfeitHolding.dateConfiscated)
          : undefined;
        weeklyCounterfeitHolding.dateSubmittedToCBK = weeklyCounterfeitHolding.dateSubmittedToCBK
          ? dayjs(weeklyCounterfeitHolding.dateSubmittedToCBK)
          : undefined;
      });
    }
    return res;
  }
}
